apps/lib/opt.c

```diff
@@ -356,21 +356,25 @@ void print_format_error(int format, unsigned long flags)
 }
 
 /* Parse a cipher name, put it in *EVP_CIPHER; return 0 on failure, else 1. */
-int opt_cipher(const char *name, const EVP_CIPHER **cipherp)
+int opt_cipher(const char *name, EVP_CIPHER **cipherp)
 {
-    *cipherp = EVP_get_cipherbyname(name);
+    *cipherp = (EVP_CIPHER *)EVP_get_cipherbyname(name);
     if (*cipherp != NULL)
         return 1;
+    *cipherp = EVP_CIPHER_fetch(NULL, name, NULL);
     opt_printf_stderr("%s: Unknown cipher: %s\n", prog, name);
     return 0;
 }
 
 /*
  * Parse message digest name, put it in *EVP_MD; return 0 on failure, else 1.
  */
-int opt_md(const char *name, const EVP_MD **mdp)
+int opt_md(const char *name, EVP_MD **mdp)
 {
-    *mdp = EVP_get_digestbyname(name);
+    *mdp = (EVP_MD *)EVP_get_digestbyname(name);
+    if (*mdp != NULL)
+        return 1;
+    *mdp = EVP_MD_fetch(NULL, name, NULL);
     if (*mdp != NULL)
         return 1;
```

Comment:
May I suggest somewhat less redundant code?

``` C
    *mdp = (EVP_MD *)EVP_get_digestbyname(name);
    if (*mdp == NULL)
       *mdp = EVP_MD_fetch(NULL, name, NULL);
    if (*mdp != NULL)
        return 1;
```

Also, just as with ciphers, perhaps fetch first, and legacy as fallback?

Reply:
I fixed the order to be fetch then legacy.
Saving one return statement, okay :)

Reply:
Even more terse (and more readable in my mind) would be:

``` C
    if ((*mdp = EVP_MD_fetch(NULL, name, NULL)) != NULL
        || (*mdp = (EVP_MD *)EVP_get_digestbyname(name)) != NULL)
        return 1;
```

Reply:
in retrospect, I'm not going to do this.  The current code is clear and obvious and squeezing a handful of bytes or CPU cycles is the wrong trade-off.